src/portfft/descriptor.hpp

```diff
@@ -491,18 +512,89 @@ class committed_descriptor {
     // Dummy parameter is needed as only partial specializations are allowed without specializing the containing class
     template <detail::level Lev, typename Dummy>
     struct inner {
-      static Scalar* execute(committed_descriptor& desc, dimension_struct& dimension_data);
+      static Scalar* execute(committed_descriptor& desc, std::vector<kernel_data_struct>& kernels);
     };
   };
 
   /**
    * Calculates twiddle factors for the implementation in use.
    *
-   * @param dimension_data data about the dimension for which twiddles are needed
+   * @param kernels vector of kernels
    * @return Scalar* USM pointer to the twiddle factors
    */
-  Scalar* calculate_twiddles(dimension_struct& dimension_data) {
-    return dispatch<calculate_twiddles_struct>(dimension_data.level, dimension_data);
+  Scalar* calculate_twiddles(detail::level level, std::vector<kernel_data_struct>& kernels) {
+    return dispatch<calculate_twiddles_struct>(level, kernels);
+  }
+
+  /**
+   * utility function to set specialization constants
+   * @tparam SubgroupSize Subgroup size
+   * @param top_level selected level of implementation
+   * @param prepared_vec vector as returned from prepare_implementation
+   * @param compute_direction direction of compute, as in forward or backward
+   * @param kernel_num which dimension are the kernels being built for
+   * @param is_compatible flag to be set if the kernels are compatible
+   * @return
+   */
+  template <Idx SubgroupSize>
+  std::vector<kernel_data_struct> set_specialization_constants(
+      detail::level top_level,
+      std::vector<std::tuple<detail::level, std::vector<sycl::kernel_id>, std::vector<Idx>>>& prepared_vec,
+      direction compute_direction, std::size_t kernel_num, bool& is_compatible) {
+    Scalar scale_factor = compute_direction == direction::FORWARD ? params.forward_scale : params.backward_scale;
+    std::size_t counter = 0;
+    bool take_conjugate_on_load = false;
+    bool take_conjugate_on_store = false;
+    std::vector<kernel_data_struct> result;
+    for (auto& [level, ids, factors] : prepared_vec) {
```

Comment:
I think taking the reference here is the right thing to do!  (`auto& [...]`). Can you also apply that in `build_w_spec_const` where lots of copying is going on?

Reply:
All that copying is avoided now. It was in that for loop itself, which is now moved to that separate function and uses `auto&`